base/server/src/main/java/com/netscape/cms/realm/PKILDAPRealm.java

```diff
@@ -0,0 +1,415 @@
+
+            } catch (LDAPException e) {
+                logger.warn("Unable to authenticate user " + user.getUserDN() + ": " + e.getMessage());
+                if (e.getLDAPResultCode() == LDAPException.INVALID_CREDENTIALS) {
+                    return null;
+                } else {
+                    throw e;
+                }
+            } finally {
+                authConn.close();
+            }
+
+            logger.info("User " + username + " authenticated");
+
+            List<String> roles = getUserRoles(conn, user.getUserDN());
+            principal = new PKIPrincipal(user, null, roles);
+
+        } catch (Exception e) {
+            logger.error("Problem to cannect the LDAP backend and verify user credentials");
+        } finally {
+            connFactory.returnConn(conn);
+        }
+        return principal;
+    }
+
+    @Override
+    public Principal authenticate(X509Certificate[] certChain) {
+        LDAPConnection conn = null;
+        PKIPrincipal principal = null;
+        try {
+            // sort cert chain from leaf to root
+            certChain = Cert.sortCertificateChain(certChain, true);
+
+            // get leaf cert
+            X509Certificate cert = certChain[0];
+            String certID = getCertID(cert);
+
+            logger.info("Authenticating user with certificate " + certID);
+
+            // cert already validated during SSL handshake
+
+            conn = connFactory.getConn();
+            logger.info("Searching for user with certificate " + certID);
+            User user = findUserByCertID(conn, certID);
+
+            if (user == null) {
+                logger.warn("Unable to authenticate user with certificate " + certID + ": User not found");
+                return null;
+            }
+
+            logger.info("Searching for matching certificates in user " + user.getUserDN());
+            X509Certificate[] certs = user.getX509Certificates();
+
+            if (certs == null || certs.length == 0) {
+                logger.warn("Unable to authenticate user " + user.getUserDN() + ": User has no certificates");
+                return null;
+            }
+
+            boolean found = false;
+            byte[] data = cert.getEncoded();
+
+            for (X509Certificate c : certs) {
+                if (Arrays.equals(data, c.getEncoded())) {
+                    found = true;
+                    break;
+                }
+            }
+
+            if (!found) {
+                logger.warn("Unable to authenticate user " + user.getUserDN() + ": No matching certificate");
+                return null;
+            }
+
+            logger.info("User " + user.getUserDN() + " authenticated");
+
+            List<String> roles = getUserRoles(conn, user.getUserDN());
+            principal = new PKIPrincipal(user, null, roles);
+
+        } catch (Exception e) {
+            logger.error("Problem to cannect the LDAP backend and verify the certificate");
```

Comment:
I think we should rethrow the exception (e.g. wrapped with `RuntimeException`) so we can troubleshoot the problem (e.g. database issue). If we only return a `null` people might think there's a problem with the cert.

Reply:
> people might think there's a problem with the cert

If we return the RuntimeException will the end user see a server error? The ACMERealm was returning an exception but I had not verified where/how it was handled. I will introduce the runtime exception and see how it works